### PR TITLE
Python bindings using PyBind11

### PR DESCRIPTION
**Pull request by mjmottram**
@jackdunger

### CODE REVIEW

**Comment by jackdunger** [comment]
@mjmottram, what's the staus of this PR?

**Comment by mjmottram** [comment]
Some improvements could be added down the line (e.g. python side oxsx -> ROOT objects) but it should be stable and OK to merge in the current state.

**Comment by jackdunger** [comment]
The python side code is definitely a separate PR. Where are we now with pybind11 on:
* Pointers
* Default arguments
* Function overloading
* Static methods
Cheers

**Comment by mjmottram** [comment]
* Pointers work out of the box, boost::python refused to compile when C++ function took a pointer, pybind11 runs fine (testing just the BinnedNLLH::SetDataSet method); have updated PR accordingly.
* Default arguments and overloaded functions work (with a small implementation detail, see e.g. BinnedPdf.wrap for implementation of both overloading and default arguments).
* Static methods work with the `def_static` method

**Comment by jackdunger** [comment]
@mjmottram sorry for the delay, I spent a long time trying to compile this on SL6 and then gave up. It works fine on my mac though

File: config/userconfig.ini.template [hunk 1]
```diff
@@ -14,3 +14,7 @@ lib_path    : /path/to/dir
 [gsl]
 header_path : /path/to/dir
 lib_path    : /path/to/dir
+
+[python]
```
Comment (jackdunger) [inline config/userconfig.ini.template]: Im not sure we need this, without python they can't run this script or `scons`. And we can get the python headers from `distutils.sysconfig` like you did for the libraries

File: config/dependencies.ini [hunk 2]
```diff
@@ -13,4 +13,8 @@ libs : Core RIO Hist Gpad Tree Rint Postscript Matrix MathCore Thread pthread m
 
 [gsl]
 check_headers : gsl/gsl_cdf.h
-libs : gsl gslcblas
\ No newline at end of file
+libs : gsl gslcblas
+
+[python]
+check_headers : python.h
```
Comment (jackdunger) [inline config/dependencies.ini]: Please see comment below

File: src/core/histogram/AxisCollection.h [hunk 3]
```diff
@@ -14,7 +14,7 @@
 class AxisCollection{
     
  public:
-    AxisCollection() : fNDimensions(0), fNBins(0){}
+    AxisCollection() : fNBins(0), fNDimensions(0){}
```
Comment (jackdunger) [inline src/core/histogram/AxisCollection.h]: ?

**Comment by jackdunger** [comment]
boring suggestion: can we call them python_bindings or pybindings or something? instead of bindings

**Comment by jackdunger** [comment]
It might have been my commit that stopped it installing on SL6, if so feel free to scrap it. Once we have it installing I'll double check all the methods are there and we can merge. 
Maybe @billyliggings could check the ubuntu build if you get it going on lunix? 🙏

**Comment by mjmottram** [comment]
@jackdunger what's the status of this at your end?  Can you get it to compile the bindings?

**Comment by mjmottram** [comment]
🔔 🔔 🔔

**Pull request closed without merging**